File: ant.rb
# Represents a single friendly ant. Enemy ants don't get a class
# TODO could we gain anything by having Ant subclass Item?
class Ant
  @@living = []
  @@next_ant_id = 0

  attr_reader :square
  attr_reader :next_square
  attr_accessor :id
  attr_accessor :goal
  attr_accessor :route

  def initialize(square)
    @square = @next_square = square
    @square.ant = self
    @square.next_ant = self
    @route = []

    @id = @@next_ant_id
    @@next_ant_id += 1

    @@living << self
  end

  def self.living
    @@living
  end

  def self.advance_turn!
    self.living.each(&:advance_turn!)
  end

  def alive?
    @@living.member?(self)
  end

  def die!
    @@living.delete(self)
    @square = @next_square = nil
  end

  # Entering this method, all squares should have @ant == nil
  def advance_turn!
    @square = @next_square
    @square.ant = @square.next_ant = self
  end

  # Order this ant to go to a given *adjacent* square and note the next expected position.
  def order_to(adjacent)
    @next_square = adjacent
    @next_square.next_ant = self

    AI.instance.order(square, square.direction_to(adjacent))
  end

  def to_s
    "<Ant #{@id} at #{@square}>"
  end
end

File: java/src/com/bradleybuda/Ant.java
package com.bradleybuda;

import java.util.HashSet;
import java.util.Set;

import org.apache.commons.lang.builder.EqualsBuilder;
import org.apache.commons.lang.builder.HashCodeBuilder;

public class Ant {
    private static final Set<Ant> LIVING = new HashSet<Ant>();
    private static int nextAntId = 0;
    
    private final int id;
    private Square square;
    private Square nextSquare;
    private Goal goal;
    
    public Ant(Square square) {
        this.id = nextAntId;
        nextAntId++;
        this.square = this.nextSquare = square;
        square.setAnt(this);
        square.setNextAnt(this);
        LIVING.add(this);
    }
    
    public static void advanceTurn() {
        for (final Ant a : LIVING)
            a.advanceTurn();
    }
    
    public boolean isAlive() {
        return LIVING.contains(this);
    }
    
    public void die() {
        LIVING.remove(this);
        square = nextSquare = null;
    }
    
    // Entering this method, all squares should have ant == null
    public void advanceTurn() {
        square = nextSquare;
        square.setAnt(this);
        square.setNextAnt(this);        
    }
    
    public void orderTo(final Square adjacent) {
        nextSquare = adjacent;
        nextSquare.setNextAnt(this);
        
        AI.order(square, square.directionTo(adjacent));
    }
    
    @Override
    public String toString() {
        return String.format("<Ant %d at %s>", id, square);
    }
    
    @Override
    public int hashCode() {
        HashCodeBuilder b = new HashCodeBuilder();
        b.append(square);
        return b.hashCode();
    }
    
    @Override
    public boolean equals(Object obj) {
        EqualsBuilder b = new EqualsBuilder();
        b.append(this.square, ((Ant)obj).square);
        return b.isEquals();
    }
}
